tensorflow/compiler/mlir/tensor/transforms/legalize_utils.cc

```diff
@@ -0,0 +1,249 @@
+
+#include "mlir/Dialect/Arith/IR/Arith.h"  // from @llvm-project
+#include "mlir/Dialect/Linalg/IR/Linalg.h"  // from @llvm-project
+#include "mlir/Dialect/Math/IR/Math.h"  // from @llvm-project
+#include "mlir/Dialect/SCF/IR/SCF.h"  // from @llvm-project
+#include "mlir/Dialect/Tensor/IR/Tensor.h"  // from @llvm-project
+#include "tensorflow/compiler/mlir/tensor/transforms/legalize_utils.h"
+
+namespace mlir {
+namespace tensor {
+
+namespace {
+
+// Return the total size of the given tensor if it is ranked and all of its
+// dimensions have static size. Otherwise return nullopt. The given value is
+// expected to be a tensor.
+std::optional<int64_t> getTensorStaticSize(Value tensor) {
+  auto tensor_type = tensor.getType().cast<TensorType>();
+  auto ranked_tensor_type = tensor_type.dyn_cast<RankedTensorType>();
+  if (!ranked_tensor_type)
+    return std::nullopt;
+
+  int64_t size = 1;
+  for (auto dim_size : ranked_tensor_type.getShape()) {
+    if (dim_size == ShapedType::kDynamic)
+      return std::nullopt;
+    size *= dim_size;
+  }
+  return size;
+}
+
+// If the given SSA value is a constant tensor of integers, return the
+// corresponding values. Otherwise, nullopt.
+std::optional<SmallVector<int64_t>> getConstantShape(Value shape) {
+  DenseIntElementsAttr shape_attr;
+  if (!matchPattern(shape, m_Constant(&shape_attr)))
+    return std::nullopt;
+
+  return llvm::map_to_vector(shape_attr.getValues<IntegerAttr>(), [](IntegerAttr attr) {
+    return attr.getInt();
+  });
+}
+
+// Return whether the given shape has a wildcard (element set to -1).
+bool constantShapeHasWildcard(ArrayRef<int64_t> shape) {
+  return llvm::any_of(shape, [](auto size) {
+    return size == -1;
+  });
+}
+
+// Substitute the occurrence of -1 in the given shape with a new positive value
+// such that the product of all values in 'input_size' is equal to the product
+// in all values of 'shape'.
+SmallVector<int64_t> substituteConstantShapeWildcard(
+    ArrayRef<int64_t> input_size, ArrayRef<int64_t> shape) {
+  auto input_size_product = std::accumulate(input_size.begin(), input_size.end(), 1, std::multiplies<int64_t>());
+  auto shape_product = std::accumulate(shape.begin(), shape.end(), 1, std::multiplies<int64_t>());
+  auto wildcard_substitution = shape_product ? input_size_product / std::abs(shape_product) : 0;
+  return llvm::map_to_vector(shape, [&](int64_t item) {
+    return item == -1 ? wildcard_substitution : item;
+  });
+}
+
+// Emit an 'arith.constant' op with the given tensor elements and type.
+Value constantShapeToValue(OpBuilder& builder, Location loc, ArrayRef<int64_t> shape, Type element_type) {
+  assert(element_type.isa<IntegerType>());
+  auto tensor_type = RankedTensorType::get({shape.size()}, element_type);
+  auto dims_attrs = llvm::map_to_vector(shape, [&](int64_t dim_size) -> Attribute {
+    return builder.getIntegerAttr(element_type, dim_size);
+  });
+  auto shape_attr = DenseIntElementsAttr::get(tensor_type, dims_attrs);
```

Comment:
https://mlir.llvm.org/doxygen/classmlir_1_1Builder.html#a7654e378a8f6e3ec0bc098303aa84a02

Reply:
The shape tensor can be of integers of any size, as given by argument `element_type`. I need to generate integer attributes of the right type individually and then create the tensor constant from them.